crates/typst-library/src/text/deco.rs

```diff
@@ -225,24 +225,79 @@ pub struct StrikeElem {
 impl Show for StrikeElem {
     #[tracing::instrument(name = "StrikeElem::show", skip_all)]
     fn show(&self, _: &mut Vt, styles: StyleChain) -> SourceResult<Content> {
+        let stroke = self.stroke(styles).unwrap_or_default();
         Ok(self.body().styled(TextElem::set_deco(Decoration {
-            line: DecoLine::Strikethrough,
-            stroke: self.stroke(styles).unwrap_or_default(),
+            // note that we do not support evade option for Strikethrough.
+            line: DecoLine::Strikethrough(stroke),
             offset: self.offset(styles),
             extent: self.extent(styles),
-            evade: false,
         })))
     }
 }
 
-/// Defines a line that is positioned over, under or on top of text.
+/// Highlight text by setting background color.
+///
+/// ## Example { #example }
+/// ```example
+/// This is #highlight[important].
+/// ```
+///
+/// Display: Highlight
+/// Category: text
+#[element(Show)]
+pub struct HighlightElem {
+    /// The color to fill in the background of the text.
+    /// (Default: Yellow)
+    ///
+    /// ```example
+    /// This is #highlight(fill: blue)[with blue]. \
+    /// ```
+    pub fill: Option<Paint>,
+
+    /// The position of the background rectangle relative to the baseline.
+    /// Read from the font tables if `{auto}`.
+    ///
+    /// ```example
+    /// #highlight(offset: -1.2em)[
+    ///   The Tale Of A Faraway background
+    /// ]
+    /// ```
+    #[resolve]
+    pub offset: Smart<Length>,
```

Comment:
Perhaps offset doesn't make the most sense here. What if we instead allowed to specify the top and bottom edge like for the text bounding box? These could default to `ascender` and `descender`. (We could also just use the normal text edges but I think it will look bad because these are intentionally quite tight.) The types would be `TopEdge` and `BottomEdge` as in `typst-library/src/text/mod.rs`.

Reply:
Thank you. I think I get the idea. I looked over the code, the top-edge and bottom-edge will be updated when in shaping as we loop over the glyphs' bounding box.

As far as I know (in many pdf or other note taking tools), the size of the highlight rectangle should be identical to how we uses cursor to select the text, so maybe a good way is to set the highlight area to have the same behaviour as the selection area. I did some random test on Safari and Chrome. It seems that the selection area differs in different application. For example, selecting text on a pdf in safari, the selection area should be from descender to ascender, no matter what the text is; 

![](https://github.com/typst/typst/assets/19356905/51598d68-013d-42f5-b54a-11b216a222f0)

but on Chrome, the selection area depends on the text itself, if it is a word like "number" then the bottom edge will be at the baseline, instead of the descender(but if we select more letters, the area will grow and bottom edge will be at the descender)

![](https://github.com/typst/typst/assets/19356905/b5bf8f5e-dd2c-4580-8678-560cb1a46e5e)

![](https://github.com/typst/typst/assets/19356905/48d7375c-2121-4b60-ac95-b1889ae116c1)

So I think maybe these are essentially two ways to determine the height of highlight: the Safari way which uses a static metric for all the texts, or the Chrome way which consider the text to be highlighted and updated by its bounding box.

My suggestion is that we can go the Chrome way and determine the top/bottom edge by the text's bounding box, and then we also add an argument similar to the curretn `extent`, but for the vertical direction, so that it can grows beyond the bounding box at the user's wish.

Reply:
With my proposed top-edge and bottom-edge approach, the user would have the choice. `top-edge: "bounds"` would be the Chrome thing and `top-edge: "ascender"` would be the Safari thing.

Reply:
Ohh yes indeed, this will work. The remaining design choice is what to be the preferable default value? @laurmaedje @reknih 

1. option1: the Safari way
top-edge: ascender
bottom-edge: descender

2. option2: the Chrome way
top-edge: bounds
bottom-edge: bounds

3. or option3: following the convention of what we have now for TextElem
top-edge: cap height
bottom-edge: baseline

Reply:
3 probably looks bad. I prefer 1 I think. more consistent if there's multiple highlights in one line.